### DISCUSSION TITLE
why is my derived/child material class requesting parameters from the parent class?

### DISCUSSION THREAD

**Discussion by batodon** [idaholab/moose#22864] [Q&A General] [answered]
Hello, 

I made two material classes; one is derived from the other (base class). I override the methods from the base class and I used different parameters in the derived class. However, when I ran an input file using the derived material class as an object in the input file, I received an error asking me to set a parameter from the parent class. The error is shown below. Could you help, pls? Thanks. 

*** ERROR ***
The parameter "a" is being retrieved before being set.

**Comment by dschwen** [comment, marked as answer]
Your derived class must call the constructor of the parent class, which will likely access parent class input parameters. Look again at the example we have for derived classes (pretty much all of the classes in framework and modules), and not how the validParams of the parent class are always added to the validParams of the child class.

**Reply by batodon** [reply]
Hello @dschwen ,
I looked through some examples of the derived classes, and I believe my derived class calls the parent constructor. I’m still getting that error. Here is a snippet example of my derived class. PEMF is the derived, and EMFBase is the base. Thanks.

```
InputParameters
PEMF::validParams()
{
  InputParameters params = EMFBase::validParams();
  params.addClassDescription(
    " This is the derived class ");    
    params.addRequiredParam<std::vector<double>>("alpha", "distance value in all 3 directions");
    params.addRequiredParam<std::vector<double>>("eps0", "threshold values");
  return params;
}

PEMF::PEMF(
    const InputParameters & parameters)
  : EMFBase(parameters),
    _alpha(getParam<std::vector<double>>("alpha")),
    _eps(getParam<std::vector<double>>("eps0"))
{}
```

**Reply by GiudGiud** [reply]
what does EMFBase look like?
this looks fine so far

**Reply by batodon** [reply]
Hi @GiudGiud,
Here is EMFBase:
```
InputParameters
EMFBase::validParams()
{
  InputParameters params = PermBase::validParams();
  params.addClassDescription(
    " This is the base");
    params.addRangeCheckedParam<Real>("a",
                                      "a > 0",
                                      " distance value");
    params.addParam<Real>("e0", "threshold value");
  return params;
}

EMFBase::EMFBase(
    const InputParameters & parameters)
  : PermBase(parameters),
    _a(getParam<Real>("a")),
    _e0(getParam<Real>("e0"))
{}
```

**Reply by GiudGiud** [reply]
and did you pass `a` in the input file?
`a` does not have a default here

**Reply by batodon** [reply]
Nope, I did not pass `a` in the input file because I’m using new parameters and methods attributed to the derived class (PEMF). I’m not sure if I have to. Do i?

**Reply by GiudGiud** [reply]
yes you have to.
The a parameter is being retrieved in the constructor for the EMFBase. The constructor for the EMFBase is called by the constructor for the derived class. The EMFBase is not written the way we intended, with this getParam that is unavoidable, the `a` parameter should have been required. Same for `e0`

**Reply by batodon** [reply]
Great! it worked after setting a default for `a` and `e0`. Thanks @GiudGiud @dschwen
